packages/payment/src/services/payment-module.ts

```diff
@@ -160,305 +176,405 @@ export default class PaymentModuleService<
+        provider_id: payment.provider_id,
+      },
+      data.amount
+    )
 
-    return await this.retrievePaymentCollection(
-      paymentCollectionId,
+    await this.refundService_.create(
       {
-        relations: ["payment_sessions"],
+        payment: data.payment_id,
+        amount: data.amount,
+        created_by: data.created_by,
       },
       sharedContext
     )
-  }
 
-  /**
-   * TODO
-   */
+    await this.paymentService_.update({ id: payment.id, data: paymentData })
 
-  authorizePaymentCollection(
-    paymentCollectionId: string,
-    sharedContext?: Context | undefined
-  ): Promise<PaymentCollectionDTO> {
-    throw new Error("Method not implemented.")
-  }
-  completePaymentCollection(
-    paymentCollectionId: string,
-    sharedContext?: Context | undefined
-  ): Promise<PaymentCollectionDTO> {
-    throw new Error("Method not implemented.")
+    return await this.retrievePayment(
+      payment.id,
+      { relations: ["refunds"] },
+      sharedContext
+    )
   }
 
-  cancelPayment(paymentId: string, sharedContext?: Context): Promise<PaymentDTO>
-  cancelPayment(
-    paymentId: string[],
-    sharedContext?: Context
-  ): Promise<PaymentDTO[]>
+  @InjectTransactionManager("baseRepository_")
+  async cancelPayment(
+    paymentId: string,
+    @MedusaContext() sharedContext?: Context
+  ): Promise<PaymentDTO> {
+    const payment = await this.paymentService_.retrieve(
+      paymentId,
+      {},
+      sharedContext
+    )
 
-  cancelPayment(
-    paymentId: string | string[],
-    sharedContext?: Context
-  ): Promise<PaymentDTO | PaymentDTO[]> {
-    throw new Error("Method not implemented.")
-  }
+    // TODO: revisit when totals are implemented
+    //   if (payment.captured_amount !== 0) {
+    //     throw new MedusaError(
+    //       MedusaError.Types.INVALID_DATA,
+    //       `Cannot cancel a payment: ${payment.id} that has been captured.`
+    //     )
+    //   }
 
-  authorizePaymentSessions(
-    paymentCollectionId: string,
-    sessionIds: string[],
-    sharedContext?: Context | undefined
-  ): Promise<PaymentCollectionDTO> {
-    throw new Error("Method not implemented.")
-  }
-  completePaymentSessions(
-    paymentCollectionId: string,
-    sessionIds: string[],
-    sharedContext?: Context | undefined
-  ): Promise<PaymentCollectionDTO> {
-    throw new Error("Method not implemented.")
+    await this.paymentProviderService_.cancelPayment({
+      data: payment.data!,
+      provider_id: payment.provider_id,
+    })
+
+    await this.paymentService_.update(
+      { id: paymentId, canceled_at: new Date() },
+      sharedContext
+    )
+
+    return this.retrievePayment(payment.id, {}, sharedContext)
   }
-  setPaymentSessions(
-    paymentCollectionId: string,
-    data: SetPaymentSessionsDTO[],
-    sharedContext?: Context | undefined
-  ): Promise<PaymentCollectionDTO> {
-    throw new Error("Method not implemented.")
+
+  private async createProvidersOnLoad() {
```

Comment:
question: should we move that to a loader like oli did for the region? you have access to the services in the container too

Reply:
comment: Yes, this will be moved to a loader in #6311